Add support for ByteString

Lerbaek opened the pull request:
This aims to fix issue #5 by supporting the `Google.Protobuf.ByteString` type in `FieldTypeResolver`.

Furthermore, the error message for unsupported field types has been improved. It now prints the name of the unsupported field type, rather than the name of the `FieldDescriptor` type.

Keep in mind that I'm not all too experienced with Protobuf, so I don't know if this fixes the entire issue or if there's more to it, but it has been verified to fix my use case.

Note also that I'm well aware of the lack of testing, but unfortunately, I'm very restricted on my work PC, and so am unable to generate anything new.

Havret commented:
Hi, and thank you for your contribution. I appreciate your efforts.

However, I've encountered an issue with the changes you've made. When I uncomment the following lines:

https://github.com/Havret/Protobuf.System.Text.Json/blob/3f9a1818492d6945ab94608e33539b92e8b67e41/test/Protobuf.System.Text.Json.Tests/Protos/simple_message.proto#L35
https://github.com/Havret/Protobuf.System.Text.Json/blob/3f9a1818492d6945ab94608e33539b92e8b67e41/test/Protobuf.System.Text.Json.Tests/SimpleMessageTests.cs#L33

I'm getting different results than expected:

The output I get differs from what's expected. Specifically, I see:  
`"bytesProperty":[97,98,99]`  
when I anticipate:  
`"bytesProperty": "YWJj"`

Furthermore, during the deserialization test, I'm getting the exception:
```
System.NotSupportedException: The collection type 'Google.Protobuf.ByteString' is abstract, an interface, or is read only, and could not be instantiated and populated. Path: $ | LineNumber: 0 | BytePositionInLine: 299.
 ---> System.NotSupportedException: The collection type 'Google.Protobuf.ByteString' is abstract, an interface, or is read only, and could not be instantiated and populated.
   --- End of inner exception stack trace ---
   at System.Text.Json.ThrowHelper.ThrowNotSupportedException(ReadStack& state, Utf8JsonReader& reader, NotSupportedException ex)
   at System.Text.Json.ThrowHelper.ThrowNotSupportedException_CannotPopulateCollection(Type type, Utf8JsonReader& reader, ReadStack& state)
   at System.Text.Json.Serialization.Converters.IEnumerableOfTConverter`2.CreateCollection(Utf8JsonReader& reader, ReadStack& state, JsonSerializerOptions options)
   at System.Text.Json.Serialization.JsonCollectionConverter`2.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, TCollection& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonResumableConverter`1.Read(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options)
   at Protobuf.System.Text.Json.InternalConverters.FieldConverter`1.Read(Utf8JsonReader& reader, IMessage obj, Type typeToConvert, JsonSerializerOptions options, IFieldAccessor fieldAccessor) in /Users/krzysztofporebski/git/Protobuf.System.Text.Json/src/Protobuf.System.Text.Json/InternalConverters/FieldConverter.cs:line 22
   at Protobuf.System.Text.Json.ProtobufConverter`1.Read(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options) in /Users/krzysztofporebski/git/Protobuf.System.Text.Json/src/Protobuf.System.Text.Json/ProtobufConverter.cs:line 110
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
```

Havret reviewed (changes requested):
Hi, and thank you for your contribution. I appreciate your efforts.

However, I've encountered an issue with the changes you've made. When I uncomment the following lines:

https://github.com/Havret/Protobuf.System.Text.Json/blob/3f9a1818492d6945ab94608e33539b92e8b67e41/test/Protobuf.System.Text.Json.Tests/Protos/simple_message.proto#L35
https://github.com/Havret/Protobuf.System.Text.Json/blob/3f9a1818492d6945ab94608e33539b92e8b67e41/test/Protobuf.System.Text.Json.Tests/SimpleMessageTests.cs#L33

I'm getting different results than expected:

The output I get differs from what's expected. Specifically, I see:  
`"bytesProperty":[97,98,99]`  
when I anticipate:  
`"bytesProperty": "YWJj"`

Furthermore, during the deserialization test, I'm getting the exception:
```
System.NotSupportedException: The collection type 'Google.Protobuf.ByteString' is abstract, an interface, or is read only, and could not be instantiated and populated. Path: $ | LineNumber: 0 | BytePositionInLine: 299.
 ---> System.NotSupportedException: The collection type 'Google.Protobuf.ByteString' is abstract, an interface, or is read only, and could not be instantiated and populated.
   --- End of inner exception stack trace ---
   at System.Text.Json.ThrowHelper.ThrowNotSupportedException(ReadStack& state, Utf8JsonReader& reader, NotSupportedException ex)
   at System.Text.Json.ThrowHelper.ThrowNotSupportedException_CannotPopulateCollection(Type type, Utf8JsonReader& reader, ReadStack& state)
   at System.Text.Json.Serialization.Converters.IEnumerableOfTConverter`2.CreateCollection(Utf8JsonReader& reader, ReadStack& state, JsonSerializerOptions options)
   at System.Text.Json.Serialization.JsonCollectionConverter`2.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, TCollection& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonResumableConverter`1.Read(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options)
   at Protobuf.System.Text.Json.InternalConverters.FieldConverter`1.Read(Utf8JsonReader& reader, IMessage obj, Type typeToConvert, JsonSerializerOptions options, IFieldAccessor fieldAccessor) in /Users/krzysztofporebski/git/Protobuf.System.Text.Json/src/Protobuf.System.Text.Json/InternalConverters/FieldConverter.cs:line 22
   at Protobuf.System.Text.Json.ProtobufConverter`1.Read(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options) in /Users/krzysztofporebski/git/Protobuf.System.Text.Json/src/Protobuf.System.Text.Json/ProtobufConverter.cs:line 110
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
```

Lerbaek commented:
Thanks for your feedback. I was aware that it was a naive approach, and the only reason it worked in my use case was probably because the ByteString was only in the contract but not in the instances I worked with.

Great to see that tests were already waiting. I've added a ByteString converter and added use of it, and now the ByteString tests pass.

I've gone with serializing to a base64 encoded string rather than the explicit byte array, as this is what both the Protobuf library and the JSON library readily supports.

Lerbaek commented on src/Protobuf.System.Text.Json/FieldTypeResolver.cs:
```diff
@@ -32,6 +33,8 @@ public static Type ResolverFieldType(FieldDescriptor fieldDescriptor, Dictionary
                 return typeof(string);
+            case FieldType.Bytes:
+                return typeof(ByteString);
             case FieldType.Message when fieldDescriptor.MessageType.ClrType is { } clrType:
```
Should we add a `ByteString` condition under the `FieldType.Message` case too?

Havret replied:
If we implement `ByteStringConverter` as `JsonConverter<ByteString?>` instead of `InternalConverter` it should work for all the cases. 

You could extend `MessageWithMaps` and corresponding `MessageWithMapsTests` with additional property:
```
map<string, bytes> map_string_to_bytes_type = 3;
```
and give it a try.

Havret reviewed (dismissed):
It works for the base case. While further work is needed to handle all cases (see [link](https://github.com/Havret/Protobuf.System.Text.Json/pull/62#discussion_r1305521407)), I believe we can merge this now and address additional concerns in separate PRs.
